code/controllers/subsystem/pathfinder.dm

```diff
@@ -1,48 +1,380 @@
+	var/turf/expand_turf		// turf we're trying to expand to
+	var/list/expand = list()	// node list of turf we're trying to expand to, if it exists.
+	var/new_cost				// new cost of a new turf being expanded to.
+	var/reverse_dir_of_expand	// reverse direction of where we're expanding to.
+	SETUP_NODE(open, null, 0, current_distance, 0, NORTH|SOUTH|EAST|WEST, start)		// initially we want to explore all cardinals.
+	while(length(open))		// while we still have open nodes
+		current = open[length(open)]		// pop a node
+		open.len--
+		current_turf = current[NODE_TURF] // get its turf
+		// see how far we are
+		CALCULATE_DISTANCE(current_turf, end)
+		// if we're at the end or close enough, we're done
+		if((current_turf == end) || (current_distance <= min_target_distance))
+			// assemble our path
+			path = list(current_turf)
+			// go up the chain
+			while(current[NODE_PREVIOUS])
+				current = current[NODE_PREVIOUS]
+				path += current[NODE_TURF]
+			// get the path in the right direction
+			reverseRange(path)
+			break		// we're done!
+		// if we get to this point, the !fun! begins.
+		if(current[NODE_DEPTH] > max_node_depth)		// too deep, skip
+			continue
+		// Run each direction
+		RUN_ASTAR(NORTH)
+		RUN_ASTAR(SOUTH)
+		RUN_ASTAR(EAST)
+		RUN_ASTAR(WEST)
+		// Clear directions, we're done with this node.
+		current[NODE_DIR] = NONE
+		CHECK_TICK
+
+	return path || PATHFIND_FAIL_NO_PATH
+
+#undef RUN_ASTAR
+
+#undef MANHATTAN
+#undef BYOND
+#undef EUCLIDEAN
+
+#undef NODE_LIST_LENGTH
+#undef NODE_PREVIOUS
+#undef NODE_WEIGHT
+#undef NODE_COST
+#undef NODE_HEURISTIC
+#undef NODE_DEPTH
+#undef NODE_DIR
+#undef SETUP_NODE
+#undef INJECTION_SETUP
+#undef INJECT_NODE
+#undef NODE
+#undef CALCULATE_DISTANCE
+
+// TURF PROCS - Should these be inlined later? Would be a loss of customization.. but uh, proccall overhead hurts!
+/**
+  * Returns whether or not a pathfinding operation with a specified caller can cross to another turf.
+  */
+/turf/proc/pathfinding_can_cross(caller, turf/other, obj/item/card/id/ID, dir_to_other, reverse_dir)
+	if(dir_to_other & (dir_to_other - 1))		// diagonal check
+		// let's handle it like how diagonal movement does realistically.
+		var/northsouth = dir_to_other & (NORTH|SOUTH)
+		var/eastwest = dir_to_other & (EAST|WEST)
+		var/turf/one = get_step(src, northsouth)
+		var/turf/two = get_step(src, eastwest)
+		return (one && pathfinding_can_cross(caller, one, ID, northsouth, REVERSE_DIR(northsouth)) && one.pathfinding_can_cross(caller, other, ID, eastwest, REVERSE_DIR(eastwest))) || (two && pathfinding_can_cross(caller, two, ID, eastwest, REVERSE_DIR(eastwest)) && two.pathfinding_can_cross(caller, other, ID, northsouth, REVERSE_DIR(northsouth)))
+	// check density first. good litmus test.
+	if(other.density)
+		return FALSE
+	// we should probably do all on edge objects but honestly can't be arsed right now
+	for(var/obj/structure/window/W in src)
```

Comment:
I didn't do this in my branch when I worked on it but holy shit this snowflakey shit for windows really should be done differently. This doesn't even work with railings atm.

Reply:
tables and grilles also defeats astar currently

Reply:
they shouldnt, they return TRUE if you have the flag PASSTABLE or PASSGRILLE

Reply:
If you have changed that, ignore me. I was referring to current on live.

Reply:
No it wasn't changed, I just remember looking at the code and seeing that tables and grilles have overrides, that, should allow ai to pass over them if the flags are set

Reply:
I have seen countless beepsky's foiled by a table or before they got shock immunity baited into grilles for the baton. If its supposed to work, its not.

Reply:
they cant pass grilles or tables, they dont have the passflags for it.

Reply:
Then why are they pathing through it?

Reply:
🤷

Reply:
these crappy for var/obj|window|doorwindow in obj loops are teh thing I was talking about with respect to a NEEDS_COMPLEX_PATH_CHECK flag on the turf